common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerOperationExecutor.java

```diff
@@ -0,0 +1,203 @@
+        @NonNull T extractResultBundle(@Nullable final Bundle resultBundle) throws BaseException;
+
+        /**
+         * Returns method name (for logging/telemetry purpose).
+         */
+        @NonNull String getMethodName();
+
+        /**
+         * ID of the telemetry API event associated to this strategy task.
+         * If this value returns null, no telemetry event will be emitted.
+         */
+        @Nullable String getTelemetryApiId();
+
+        /**
+         * A method that will be invoked before the success event is emitted.
+         * If the calling operation wants to put any value in the success event, put it here.
+         */
+        void putValueInSuccessEvent(ApiEndEvent event, T result);
+    }
+
+    private final List<IIpcStrategy> mStrategies;
+
+    /**
+     * @param strategies list of IIpcStrategy to be invoked.
+     */
+    public BrokerOperationExecutor(@NonNull final List<IIpcStrategy> strategies) {
+        mStrategies = strategies;
+    }
+
+    /**
+     * A generic method that would initialize and iterate through available strategies.
+     * It will return a result immediately if any of the strategy succeeds, or throw an exception if all of the strategies fails.
+     */
+    public <T extends CommandParameters, U> U execute(@NonNull final T parameters,
+                                                      @NonNull final BrokerOperation<U> operation)
+            throws BaseException {
+        final String methodName = ":execute";
+
+        emitOperationStartEvent(parameters, operation);
+
+        if (mStrategies.size() == 0) {
+            final ClientException exception = new ClientException(
+                    ErrorStrings.BROKER_BIND_SERVICE_FAILED,
+                    "No strategies can be used to connect to the broker.");
+            emitOperationFailureEvent(operation, exception);
+            throw exception;
+        }
+
+        final List<BrokerCommunicationException> communicationExceptionStack = new ArrayList<>();
+        for (final IIpcStrategy strategy : mStrategies) {
+            try {
+                final U result = performStrategy(strategy, operation);
+                emitOperationSuccessEvent(operation, result);
+                return result;
+            } catch (final BrokerCommunicationException communicationException) {
+                // Fails to communicate to the . Try next strategy.
+                communicationExceptionStack.add(communicationException);
+            } catch (final BaseException exception) {
+                emitOperationFailureEvent((BrokerOperation<U>) operation, exception);
+                throw exception;
+            }
+        }
+
+        // This means that we've tried every strategies... log everything...
+        for (final BrokerCommunicationException e : communicationExceptionStack) {
+            Logger.error(TAG + methodName, e.getMessage(), e);
+        }
+        final ClientException exception = new ClientException(
```

Comment:
We have API 19+, right?  We can add all the exceptions that you caught in this processed as suppressedExceptions to the exception that you throw by calling addSuppressed(e) on it.

Reply:
min is 16. I could add this 

if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                exception.addSuppressed(e);
            }